File: LeetCode/[M] - 1372. Longest ZigZag Path in a Binary Tree/solution.py
from typing import Optional


# TreeNode - Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.res = 0

    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.searchZigZag(root.left, "r", 1)
        self.searchZigZag(root.right, "l", 1)
        return self.res

    def searchZigZag(self, node, direction, curr):
        if not node:
            return
        if direction == "r":
            self.searchZigZag(node.left, "r", 1)
            self.searchZigZag(node.right, "l", curr + 1)
        elif direction == "l":
            self.searchZigZag(node.left, "r", curr + 1)
            self.searchZigZag(node.right, "l", 1)
        self.res = max(self.res, curr)

File: LeetCode/[M] - 1372. Longest ZigZag Path in a Binary Tree/solution.go
package main

// TreeNode - Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func longestZigZag(root *TreeNode) int {
	var res int
	searchZigZag(root.Right, 'l', 1, &res)
	searchZigZag(root.Left, 'r', 1, &res)
	return res
}

func searchZigZag(node *TreeNode, direction byte, curr int, res *int) {
	if node == nil {
		return
	}
	if direction == 'r' {
		searchZigZag(node.Right, 'l', curr+1, res)
		searchZigZag(node.Left, 'r', 1, res)
	} else if direction == 'l' {
		searchZigZag(node.Left, 'r', curr+1, res)
		searchZigZag(node.Right, 'l', 1, res)
	}
	*res = max(*res, curr)
}
